In [1]:
import numpy as np
from hw8_dataload import LFD_Data2

# FINAL

## Nonlinear Transforms

A polynomial transform of order Q = 10 applied to &Xscr; of dimension *d* = 2 results in a &Zscr; space of what dimensionality?

Note that this number is a triangular number. z<sub>n</sub> be the dimension of the order n transform and let's write down a few examples and a and b be the coordinates in the &Xscr; space. Also, let T<sub>n</sub> be the nth triangular number (with T<sub>0</sub> = 0) and P<sub>n</sub> be the set of terms resulting from the polynomial transform of order n.


P<sub>2</sub> &rArr; {1, a, b, a<sup>2</sup>b<sup>0</sup>, ab, a<sup>0</sup>b<sup>2</sup>} &rArr; z<sub>2</sub> = 6 = T<sub>3</sub>

P<sub>3</sub> &rArr; P<sub>2</sub> &Union; {a<sup>3</sup>b<sup>0</sup>, a<sup>2</sup>b<sup>1</sup>, a<sup>1</sup>b<sup>2</sup>, a<sup>0</sup>b<sup>3</sup>} &rArr; z<sub>2</sub> + 4 = 10 = T<sub>4</sub>

Thus, z<sub>n</sub> = T<sub>n+1</sub> and T<sub>11</sub> = z<sub>10</sub> = 55.

## Bias and Variance

Give average hypothesis g-bar resulting from training the same model &Hscr; on different data sets &Dscr; to get g<sup>(&Dscr;)</sup> &isin; &Hscr; and taking the expected value of g<sup>(&Dscr;)</sup> w.r.t. &Dscr;. We want to consider a model &Hscr; that could result in g-bar &notin; &Hscr;.

As the logistic regression function &theta;(s) = e<sup>s</sup>/(1+e<sup>s</sup>) is non-linear, we can expect the averaging done by taking the expected value not to be in &Hscr;.

## Overfitting

We **cannot** always determine if there is overfitting by comparing the values of (E<sub>out</sub> - E<sub>in</sub>). One major issue is that we have **no** idea of what E<sub>in</sub> is doing. (E<sub>out</sub> - E<sub>in</sub>) = 0.9 - 0.5 = 0.5 -0.1 = 0.4. In the first case, it's arguable that our hypothesis doesn't really match the target function all that well, it's just bad yet our difference vaue is the same as the second case, where our hypothesis is doing relatively well and more of a candidate of overfitting.

Deterministic noise is dependent on the hypothesis set since it's the details of the target function f that it cannot possibly capture. Stochastic noise however is not dependent on the hypthesis set, it's just regular old noise in the data.

## Regularization

**Given :** The regularized weight w<sub>reg</sub> solves:

minimize (1/N)&sum;(n=1;N){w<sup>T</sup>x<sub>n</sub>-y<sub>n</sub>)<sup>2</sup> subject to w<sup>T</sup>&Gamma;<sup>T</sup>&Gamma;w &le; C

where &Gamma; is a  matrix. 

**Want :** w<sup>T</sup><sub>lin</sub>&Gamma;<sup>T</sup>&Gamma;w<sub>lin</sub> &rArr; w<sub>reg</sub> = ?

Since we wish to find the w (w<sub>reg</sub>) that minimizes the given expression and w<sub>lin</sub> already satisfies the given constrants, w<sub>reg</sub> = w<sub>lin</sub>.

Furthermore, minimizing the given function results with the given constraints results in minimizing the in-sample error function with an additional slack term, which can be considered an equation for **augmented error**.

## Regularized Linear Regression

Using the dataset of HW8 (US Postal Service Zip Code data set) and implementing the same classifiers as before (one-vs-one with one number being +1, the other being -1, rest ignored, and one-vs-all with one number being +1, the rest -1), we wish to use regularized linear regression this time, **minimizing** the following expression:


(1/N) &sum;(n=1;N){w<sup>T</sup>z<sub>n</sub>-y<sub>n</sub>)<sup>2</sup> + (&lambda;/N)w<sup>T</sup>w

On x-vs-all classifiers, we let &lambda; = 1 and both omit (z = x = (1,x<sub>1</sub>,x<sub>2</sub>)) and apply a feature transform (z = (1,x<sub>1</sub>,x<sub>2</sub>,x<sub>1</sub>x<sub>2</sub>,x<sub>1</sub><sup>2</sup>,x<sub>2</sub><sup>2</sup>))

Then we want to use a 1-vs-5 classifier with the feature tranform and compare the performance with &lambda; = 0.01 and &lambda; = 1.


**Note** that minimizing the given function above is equivalent to solving for w<sub>reg</sub> = (Z<sup>T</sup>Z + &lambda;I)<sup>-1</sup>Z<sup>T</sup>y




In [2]:
#data loader
hw8_train = "hw8_train.dta"
hw8_test = "hw8_test.dta"
hw8_data = LFD_Data2(hw8_train, hw8_test)

In [3]:
#init reglinreg

from reglinreg import RegLinReg

#argts are l_reg, nlt
rlr = RegLinReg(1, False) #no feature transform

In [9]:
for cur_num in range(10):
    hw8_data.set_filter([cur_num]) #setting to x-vs-all
    #some data loading
    rlr_xtrain = hw8_data.get_X("train")
    rlr_ytrain= hw8_data.get_Y("train")
    rlr_xtest = hw8_data.get_X("test")
    rlr_ytest= hw8_data.get_Y("test")
    rlr.set_nlt(False) #no feature transform
    rlr.train_reg(rlr_xtrain, rlr_ytrain)
    rlr_ein = rlr.calc_error(rlr_xtrain, rlr_ytrain)
    rlr_eout = rlr.calc_error(rlr_xtest, rlr_ytest)
    print("~~ %d-vs-all ~~" % cur_num)
    print("no transform: e_in = %f, e_out = %f" % (rlr_ein, rlr_eout))
    rlr.set_nlt(True) #yes feature transform
    rlr.train_reg(rlr_xtrain, rlr_ytrain)
    rlr_ein = rlr.calc_error(rlr_xtrain, rlr_ytrain)
    rlr_eout = rlr.calc_error(rlr_xtest, rlr_ytest)
    print("transform: e_in = %f, e_out = %f" % (rlr_ein, rlr_eout))
    print("")

~~ 0-vs-all ~~
no transform: e_in = 0.109313, e_out = 0.115097
transform: e_in = 0.102318, e_out = 0.106627

~~ 1-vs-all ~~
no transform: e_in = 0.015224, e_out = 0.022422
transform: e_in = 0.012344, e_out = 0.021923

~~ 2-vs-all ~~
no transform: e_in = 0.100261, e_out = 0.098655
transform: e_in = 0.100261, e_out = 0.098655

~~ 3-vs-all ~~
no transform: e_in = 0.090248, e_out = 0.082711
transform: e_in = 0.090248, e_out = 0.082711

~~ 4-vs-all ~~
no transform: e_in = 0.089425, e_out = 0.099651
transform: e_in = 0.089425, e_out = 0.099651

~~ 5-vs-all ~~
no transform: e_in = 0.076258, e_out = 0.079721
transform: e_in = 0.076258, e_out = 0.079223

~~ 6-vs-all ~~
no transform: e_in = 0.091071, e_out = 0.084704
transform: e_in = 0.091071, e_out = 0.084704

~~ 7-vs-all ~~
no transform: e_in = 0.088465, e_out = 0.073244
transform: e_in = 0.088465, e_out = 0.073244

~~ 8-vs-all ~~
no transform: e_in = 0.074338, e_out = 0.082711
transform: e_in = 0.074338, e_out = 0.082711

~~ 9-vs-all ~~
no t

For the no-transform case, out of (5-9)-vs-all, 8 has the lowest E<sub>in</sub>.

For the transformed case, out of (0-4)-vs-all, 1 has the lowest E<sub>out</sub>.'

The transform improved the out-of-sample performance of 5-vs-all (7.9721% no-transform vs 7.9223% transformed) but by less than 5% (much less, closer to 0.05%).

In [12]:
hw8_data.set_filter([1, 5]) #setting to 1-vs-5
#some data loading
rlr_xtrain = hw8_data.get_X("train")
rlr_ytrain= hw8_data.get_Y("train")
rlr_xtest = hw8_data.get_X("test")
rlr_ytest= hw8_data.get_Y("test")
rlr.set_nlt(True) #feature transform

#possible lambda regularizers
l_reg = [0.01, 1]

for cur_lreg in l_reg:
    rlr.set_lambda(cur_lreg)
    rlr.train_reg(rlr_xtrain, rlr_ytrain)
    rlr_ein = rlr.calc_error(rlr_xtrain, rlr_ytrain)
    rlr_eout = rlr.calc_error(rlr_xtest, rlr_ytest)
    print("lambda = %f" % cur_lreg)
    print("e_in = %f, e_out = %f" % (rlr_ein, rlr_eout))
    print("")
    
    
    

lambda = 0.010000
e_in = 0.004484, e_out = 0.028302

lambda = 1.000000
e_in = 0.005125, e_out = 0.025943



Since from &lambda; = 0.01 to &lambda; = 1, E<sub>in</sub> increases while E<sub>out</sub> decreases, it is clear that the former case exhibits overfitting. 

## Support Vector Machines

**Given :** training set generated by target function f: &Xscr; &rarr; {-1,+1} where &Xscr; = &reals;<sup>2</sup>.


x<sub>1</sub>(1,0) &rarr; -1

x<sub>2</sub>(0,1) &rarr; -1

x<sub>3</sub>(0,-1) &rarr; -1

x<sub>4</sub>(-1,0) &rarr; +1

x<sub>5</sub>(0,2) &rarr; +1

x<sub>6</sub>(0,-2) &rarr; +1

x<sub>7</sub>(-2,0) &rarr; +1


Mapping into two-dimension space &Zscr; with:

z<sub>1</sub> = x<sub>2</sub><sup>2</sup> - 2x<sub>1</sub> - 1

z<sub>2</sub> = x<sub>1</sub><sup>2</sup> - 2x<sub>2</sub> + 1


**Want :** w,b &SuchThat; w<sup>T</sup>z + b = 0 maximizes the margin of the &Zscr; space.

In [2]:
#calculating the X stransform and setting up arrays.

fsvm_x = np.array([[1,0],[0,1], [0,-1], [-1,0], [0,2],[0,-2], [-2,0]])
fsvm_y = np.array([-1,-1,-1,1,1,1,1])
fsvm_z = np.ndarray((0,2))
for x in fsvm_x:
    z1 = pow(x[1],2.0)-(2.0*x[0])-1.0
    z2 = pow(x[0],2.0)-(2.0*x[1])+1.0
    z = np.array([z1,z2])
    fsvm_z = np.vstack((fsvm_z,z))

print("The transformed set is: ")
print(fsvm_z)

The transformed set is: 
[[-3.  2.]
 [ 0. -1.]
 [ 0.  3.]
 [ 1.  2.]
 [ 3. -3.]
 [ 3.  5.]
 [ 3.  5.]]


In [25]:
#for the heck of it, let's do this by brute force
#and go through all the given options to find the values of w and b

#our candidates
fsvm_wopt = np.array([[-0.5,-1,1],[-0.5,1,-1],[-0.5,1,0],[-0.5,0,1]])
fsvm_zt = np.c_[np.ones(fsvm_z.shape[0]),fsvm_z]

fsvm_yt = np.ndarray((0,fsvm_y.shape[0]))

for w in fsvm_wopt:
    cur_y = np.matmul(fsvm_zt, w.T)
    fsvm_yt = np.vstack((fsvm_yt, cur_y))
#multiplying by proper labels fsvm_y because we
#want the given params to actually AGREE with the output of the separating plane 
fsvm_ydist = np.sum(np.multiply(fsvm_y, fsvm_yt), axis=1)

print("\nthe outputs of candidate w plugged into the plane eq are:")
print(fsvm_yt)
print("\nmargins of candidates w (a,b,c,d):")
print(fsvm_ydist)


the outputs of candidate w plugged into the plane eq are:
[[ 4.5 -1.5  2.5  0.5 -6.5  1.5  1.5]
 [-5.5  0.5 -3.5 -1.5  5.5 -2.5 -2.5]
 [-3.5 -0.5 -0.5  0.5  2.5  2.5  2.5]
 [ 1.5 -1.5  2.5  1.5 -3.5  4.5  4.5]]

margins of candidates w (a,b,c,d):
[ -8.5   7.5  12.5   4.5]


From the preceding output, we see that w<sub>1</sub>, w<sub>2</sub>, b = 1,0,-0.5 gives us the maximal margin.

In [5]:
#hard margin svm -> C = 0
my_svm = SVM_Poly(2,0) #degree 2 polynomial kernel with hard-margin
my_svm.train(fsvm_x, fsvm_y)
cur_numalphas = my_svm.num_alphas
print("number of support vector for svm with degree 2 polynomial kernel: %d" % cur_numalphas)

number of support vector for svm with degree 2 polynomial kernel: 5


Using a hard-margin SVM with the kernel:

K(x,x') = (1+x<sup>T</sup>x')<sup>2</sup>

We end up with an SVM with 5 supoort vectors.


## Radial Basis Functions

Experimenting wtih the RBF model **a)** in regular form (Lloyd's algorithm + psuedo-inverse) with K centers:


sign(&sum; (k=1;K){w<sub>k</sub> exp(-&gamma; &Vert;x-&mu;<sub>k</sub>&Vert;<sup>2</sup>) + b)


and using w = (&Phi;<sup>T</sup>&Phi;)<sup>-1</sup>&Phi;<sup>T</sup>y, where &Phi; is the matrix made up the given expression above (or basically we want the psuedo-inverse of &Phi; times y)


and in **b)** kernel form using a hard-margin SVM:


sign(&sum; (&alpha;<sub>n</sub> &gt; 0 ){&alpha;<sub>n</sub>y<sub>n</sub> exp(-&gamma; &Vert;x-x;<sub>n</sub>&Vert;<sup>2</sup>) + b)

**Given :** target func f(x) = sign(x<sub>2</sub>-x<sub>1</sub> + 0.25(sin(&pi;x<sub>1</sub>))

In [2]:
#class for generating dataset

class RbfFData:
    def __init__(self, n):
        #n = number of data points
        n = max(1, n)
        X = np.random.uniform(-1,1, (n, 2))
        X_sin = np.sin(np.multiply(np.pi, X[:,0]))
        X_oper = np.array([-1, 1, 0.25])
        X_res = np.c_[X, X_sin]
        self.X = X
        self.Y = np.sign(np.matmul(X_res, X_oper))
        
#importing my SVM_RBF kernel class
from svm_kernel import SVM_RBF
from rbf import RBF

In [3]:
n_train = 100
n_runs = 200
gamma = [1.5, 2]
thresh = 1.0e-100
my_svm = SVM_RBF(gamma[0])
rbf_k = [9, 12]
rbf_rng = [-1,1]

#gamma, k
ein_15_9 = np.array([])
ein_15_12 = np.array([])
ein_2_9 = np.array([])
eout_15_9 = np.array([])
eout_15_12 = np.array([])
eout_2_9 = np.array([])

#kernel beats regular in eout
svm_beats_rbf9 = np.array([])
svm_beats_rbf12 = np.array([])

#svm nonseparable (ein not 0)
svm_nonsep = np.array([])

#ein is 0 for rbf, gamma 1.5, k = 9
rbf9_ein0 = np.array([])

for run in range(n_runs):
    cur_train = RbfFData(n_train)
    cur_test = RbfFData(n_train)
    my_rbf = RBF(gamma[0], cur_train.X, cur_train.Y, rbf_k[0], rbf_rng)
    my_svm.train(cur_train.X, cur_train.Y)
    svm_ein = my_svm.calc_error(cur_train.X, cur_train.Y)
    svm_eout = my_svm.calc_error(cur_test.X, cur_test.Y)
    svm_nonsep = np.concatenate((svm_nonsep, [svm_ein > thresh]))
    for k in rbf_k:
        for g in gamma:
            my_rbf.set_k(k)
            my_rbf.set_gamma(g)
            my_rbf.train()
            if k == 9 and g == 1.5:
                cein = my_rbf.calc_error(cur_train.X, cur_train.Y)
                cout = my_rbf.calc_error(cur_test.X, cur_test.Y)
                ein_15_9 = np.concatenate((ein_15_9, [cein]))
                eout_15_9 = np.concatenate((eout_15_9, [cout]))
                svm_beats_rbf9 = np.concatenate((svm_beats_rbf9, [svm_eout < cout]))
            elif k == 12 and g == 1.5:
                cein = my_rbf.calc_error(cur_train.X, cur_train.Y)
                cout = my_rbf.calc_error(cur_test.X, cur_test.Y)
                ein_15_12 = np.concatenate((ein_15_12, [cein]))
                eout_15_12 = np.concatenate((eout_15_12, [cout]))
                svm_beats_rbf12 = np.concatenate((svm_beats_rbf12, [svm_eout < cout]))
            elif k == 9 and g == 2:
                cein = my_rbf.calc_error(cur_train.X, cur_train.Y)
                cout = my_rbf.calc_error(cur_test.X, cur_test.Y)
                ein_2_9 = np.concatenate((ein_2_9, [cein]))
                eout_2_9 = np.concatenate((eout_2_9, [cout]))
            
pct_svm_nonsep =  100.0 * np.sum(svm_nonsep)/float(n_runs)
pct_rbf9_loses = 100.0 * np.sum(svm_beats_rbf9)/float(n_runs)
pct_rbf12_loses = 100.0 * np.sum(svm_beats_rbf12)/float(n_runs)
pct_rbf9_ein0 = 100.0 * np.sum(rbf9_ein0)/float(n_runs)
avg_ein_15_9 = 100.0 * np.average(ein_15_9)
avg_eout_15_9 = 100.0 * np.average(eout_15_9)
avg_ein_15_12 = 100.0 * np.average(ein_15_12)
avg_eout_15_12 = 100.0 * np.average(eout_15_12)
avg_ein_2_9 = 100.0 * np.average(ein_2_9)
avg_eout_2_9 = 100.0 *np.average(eout_2_9)

print("gamma = 1.5 | SVM non-separable pct: %f%%" % pct_svm_nonsep)
print("gamma = 1.5, k = 9 | SVM beats reg RBF pct: %f%%" % pct_rbf9_loses)
print("gamma = 1.5, k = 12 | SVM beats reg RBF pct: %f%%" % pct_rbf12_loses)
print("gamma = 1.5, k = 9 | reg RBF E_in = 0 pct: %f%%" % pct_rbf9_ein0)
print("")
print("~~gamma = 1.5 k=9~~")
print("E_in: %f%%" % avg_ein_15_9)
print("E_out: %f%%" % avg_eout_15_9)
print("")
print("~~gamma = 1.5 k=12~~")
print("E_in: %f%%" % avg_ein_15_12)
print("E_out: %f%%" % avg_eout_15_12)
print("")
print("~~gamma = 2 k=9~~")
print("E_in: %f%%" % avg_ein_2_9)
print("E_out: %f%%" % avg_eout_2_9)

gamma = 1.5 | SVM non-separable pct: 0.000000%
gamma = 1.5, k = 9 | SVM beats reg RBF pct: 85.500000%
gamma = 1.5, k = 12 | SVM beats reg RBF pct: 78.500000%
gamma = 1.5, k = 9 | reg RBF E_in = 0 pct: 0.000000%

~~gamma = 1.5 k=9~~
E_in: 4.975000%
E_out: 6.825000%

~~gamma = 1.5 k=12~~
E_in: 3.630000%
E_out: 6.080000%

~~gamma = 2 k=9~~
E_in: 5.370000%
E_out: 7.200000%


From the above, we see that as k increases, both E<sub>in</sub> and E<sub>out</sub> drop and when &gamma; increases, both E<sub>in</sub> and E<sub>out</sub> increase.